### ISSUE TITLE
[SLPVectorizer] codegen regression

### ISSUE THREAD

**Issue by davidbolvansky** [llvm/llvm-project#54985]
```
unsigned char ur[16], ua[16], ub[16];

void avgu (void)
{
  int i;

  for (i = 0; i < 16; i++)
    ur[i] = (ua[i] + ub[i] + 1) >> 1;
}
```

LLVM 11 -O3
```
avgu():                               # @avgu()
        movdqa  xmm0, xmmword ptr [rip + ub]
        pavgb   xmm0, xmmword ptr [rip + ua]
        movdqa  xmmword ptr [rip + ur], xmm0
        ret
```

LLVM trunk -O3
```
avgu():                               # @avgu()
        movq    xmm0, qword ptr [rip + ua]      # xmm0 = mem[0],zero
        movq    xmm1, qword ptr [rip + ub]      # xmm1 = mem[0],zero
        pavgb   xmm1, xmm0
        movq    xmm0, qword ptr [rip + ua+8]    # xmm0 = mem[0],zero
        movq    xmm2, qword ptr [rip + ub+8]    # xmm2 = mem[0],zero
        pavgb   xmm2, xmm0
        punpcklqdq      xmm1, xmm2              # xmm1 = xmm1[0],xmm2[0]
        movdqa  xmmword ptr [rip + ur], xmm1
        ret
```

LLVM trunk -Os looks fine indeed..
```
avgu():                               # @avgu()
        movdqa  xmm0, xmmword ptr [rip + ua]
        pavgb   xmm0, xmmword ptr [rip + ub]
        movdqa  xmmword ptr [rip + ur], xmm0
        ret
```

https://godbolt.org/z/cbKa3zT4M

**Comment by davidbolvansky** [comment]
But the IR is worse with -O3 pipeline... so seems like a midend issue..

**Comment by davidbolvansky** [comment]
cc @rotateright @fhahn @RKSimon

**Comment by topperc** [comment]
Looks like trunk -O3 is emitting two 8xi8 operations. The backend is glueing the stores together leading to the punpck and 128-bit store.

**Comment by RKSimon** [comment]
This might be a similar problem to #48223

**Comment by RKSimon** [comment]
> This might be a similar problem to #48223

Nope - its SLP not wanting to generate 256-bit vectors on pre-AVX targets

**Comment by davidbolvansky** [comment]
Yeah, this is a general regression; 

ARM case:
https://godbolt.org/z/45bdq4jsT

Please feel free to update the title with better problem description.

**Comment by davidbolvansky** [comment]
Seems like for -Os case, LoopVectorizer did the right job; for -O3 case, the loop is fully unrolled and SLPVectorizer kicks in, but not optimally.

@alexey-bataev @anton-afanasyev @fhahn

**Comment by davidbolvansky** [comment]
cc @sjoerdmeijer

**Comment by sjoerdmeijer** [comment]
FWIW: this doesn't seem X86 specific, the ARM case shows similar problems. I have looked only very quickly, but looks like what Craig said, the loop vectoriser creates only <8 x i8>s, I guess the exts must be getting in the way...

**Comment by alexey-bataev** [comment]
SLP has the analysis for the sizes of the vectors. Here we have `store i8 (trunc i16 to i8)` ops. Because of `trunc` ops the max reg size is set to 16 and max VF is defined as `128 (Max SSE vec reg size) / 16 = 8`. Because of this we try just to vectorize 8xstores and again 8xstores.

**Comment by davidbolvansky** [comment]
Not familiar with vectorizers, but it sounds strange to me than SLPVectorizer made a different decision than LoopVectorizer.

**Comment by alexey-bataev** [comment]
Main reason - compile time. We may spend some extra time for unsuccessfull attempts.

**Comment by davidbolvansky** [comment]
So then another motivation for https://reviews.llvm.org/D102748

Leave loops to loopvectorizer…

**Comment by alexey-bataev** [comment]
https://reviews.llvm.org/D92824 triggered the regression

**Comment by alexey-bataev** [comment]
Candidate patch https://reviews.llvm.org/D124284

**Comment by fhahn** [comment]
@alexey-bataev It looks like the linked review landed in 9dc4ced204d1e918d4c8d3279e52197e9a5abc94. Can the issue be closed?

**Comment by alexey-bataev** [comment]
Yep.
